### PR TITLE
Refactoring of FindGridSpots and ancillary functions

### PR DESCRIPTION
**Pull request by effting**
Goals:
- improving readability
- improving robustness
- improving test coverage
- bug-fix in handedness

Includes a module to work with graphs (structures made of vertices and edges).

This PR does not deprecate `FindGridSpots` as it is still in use by `/scripts/spot-grid.py` as well as multiple calibrations in `fastem-calibrations`

### CODE REVIEW

File: src/odemis/util/registration.py [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+
+    """
+    tform = _initial_estimate_mpp_orientation(xy)
+    grid = unit_gridpoints(shape, mode="xy")
+    correspondences = bijective_matching(tform.apply(grid), xy)
+    a, b = zip(*correspondences)
+    return transform_type.from_pointset(grid[a, ...], xy[b, ...])
+
+
+def estimate_mpp_orientation_from_img(
```
Comment (tepals) [inline src/odemis/util/registration.py:413]: I would rename this to `estimate_grid_orientation_from_img`

**Review by tepals — CHANGES_REQUESTED** [review]

File: src/odemis/util/__init__.py [hunk 2]
```diff
@@ -19,33 +19,220 @@
 You should have received a copy of the GNU General Public License along with
 Odemis. If not, see http://www.gnu.org/licenses/.
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/__init__.py]: I moved those functions to `odemis.util.random`

File: src/odemis/util/__init__.py [hunk 3]
```diff
@@ -19,33 +19,220 @@
...
 
+try:
+    from numpy.random import Generator as Generator
+except ImportError:
+
+    class Generator:  # type: ignore[no-redef]
+        pass
+
+
+T = TypeVar("T")
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/__init__.py:52]: done

File: src/odemis/util/__init__.py [hunk 4]
```diff
@@ -19,33 +19,220 @@
...
+
 
 # helper functions
+def pairwise(iterable: Iterable[T]) -> Iterable[Tuple[T, T]]:
+    "s -> (s0,s1), (s1,s2), (s2, s3), ..."
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/__init__.py]: done

File: src/odemis/util/__init__.py [hunk 5]
```diff
@@ -19,33 +19,220 @@
...
+    raise ValueError(
+        "%r cannot be used to seed a numpy.random.RandomState instance" % seed
+    )
+
+
+def rng_integers(gen, low, high=None, size=None, dtype="int64", endpoint=False):
+    """
+    Return random integers from low (inclusive) to high (exclusive), or if
+    endpoint=True, low (inclusive) to high (inclusive). Replaces
+    `RandomState.randint` (with endpoint=False) and
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/__init__.py]: Updated docstring - done

File: src/odemis/util/graph.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+        """
+
+        def __getitem__(self, i):
+            if isinstance(i, slice):
+                return self.__class__(self.data[i])
+            else:
+                return self.data[i]
+
+
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/graph.py:50]: I agree it looks a bit weird but this is how it's supposed to be according to PEP 8 / auto-formatting via `black`. I prefer to keep it this way

File: src/odemis/util/graph.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+else:
+    from collections import UserList
+
+
+class GraphBase(UserList, metaclass=ABCMeta):
+    """Abstract base class for graphs."""
+
+    _item_type = object
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/graph.py:58]: `_item_type` is to be overridden by the actual implementation. I put the most generic factory method here that I could think of.

File: src/odemis/util/graph.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        # largest edge weight (distance).
+        for triangle in triangles:
+            edges = list(itertools.combinations(triangle, 2))
+            selected = max(edges, key=self.get_edge_weight)
+            self.remove_edge(selected)
+
+
+class WeightedGraph(GraphBase):
+    """
+    represented as a list of dicts.
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/graph.py]: Updated docstring

File: src/odemis/util/graph.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            del self.data[i][j]
+
+    def get_edge_weight(self, edge: Tuple[int, int]) -> float:
+        j, i = edge
+        return self.data[j][i]
+
+
+class UnweightedGraph(GraphBase):
+    """
+    represented as a list of sets.
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/graph.py]: Updated docstring

File: src/odemis/util/graph.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+            self.data[i].add(j)
+
+    def remove_edge(self, edge: Tuple[int, int]) -> None:
+        j, i = edge
+        self.data[j].remove(i)
+        if not self._directed:
+            self.data[i].remove(j)
+
+    def get_edge_weight(self, edge: Tuple[int, int]) -> float:
+        return 1
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/graph.py:275]: Done

File: src/odemis/util/registration.py [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+    Recognition Letters, 28(12), 1523-1533.
+
+    """
+    dm = scipy.spatial.distance.cdist(src, dst, "euclidean")
+    yield from _bijective_matching(dm)
+
+
+def unit_gridpoints(shape: Tuple[int, int], *, mode: str) -> numpy.ndarray:
+    """
+    Returns an ordered array of coordinates of a square grid with unit spacing.
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/registration.py]: Done

File: src/odemis/util/registration.py [hunk 12]
```diff
@@ -0,0 +1,456 @@
...
+        for neighbor, distance in zip(neighbors, distances[vertex]):
+            # An edge connects two vertices that are each others nearest neighbor.
+            if (vertex < neighbor) and (vertex in indices[neighbor]):
+                graph.add_edge((vertex, neighbor), distance)
+
+    # The code above assumes that each point has 4 nearest neighbors. In
+    # practice we would like to only consider the 2 nearest neighbors for the
+    # points at the corners of the grid, and likewise 3 nearest neighbors for
+    # the points on the sides of the grid. The following line takes care of
+    # that.
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/registration.py:237]: Updated docstring - done

File: src/odemis/util/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/__init__.py]: done

File: src/odemis/util/registration.py [hunk 13]
```diff
@@ -0,0 +1,456 @@
...
+    1: numpy.array([(0, 1), (-1, 0)]),
+    2: numpy.array([(-1, 0), (0, -1)]),
+}
+
+
+def _canonical_matrix_form(matrix: numpy.ndarray) -> numpy.ndarray:
+    """
+    Returns a transformation matrix in canonical form.
+
+    The multi-probe pattern has a dual mirror symmetry as well as a 4-fold
```
[the remark this answers is not in the document]
Comment (tepals) [reply src/odemis/util/registration.py:256]: I would go for the second option. For now most of the other code has been written to work for rectangular and square grids. So I think it makes sense to raise a ValueError for non-square grids, but I think we wont be using non-square grids in the near future.

[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/registration.py:256]: `estimate_grid_orientation` and `estimate_grid_orientation_from_img` now raise a `NotImlementedError`

File: src/odemis/util/registration.py [hunk 14]
```diff
@@ -0,0 +1,456 @@
...
+            probs = D2 / D2.sum()
+            cumprobs = probs.cumsum()
+            r = rng.uniform()
+            init[i, :] = data[numpy.searchsorted(cumprobs, r)]
+    return init
+
+
+def _initial_estimate_mpp_orientation(xy: numpy.ndarray) -> AffineTransform:
+    """
+    Provide an initial estimate for the orientation of a square grid of points.
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/registration.py:340]: It does not work for a rectangular grid.

File: src/odemis/util/registration.py (hunk 1 above)
Comment (effting) [reply src/odemis/util/registration.py]: done

File: src/odemis/util/test/registration_test.py [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+                with self.subTest(shape=shape, i=i):
+                    n, m = shape
+                    shuffle = self._rng.permutation(n * m)
+                    tform = random_affine_transform(self._rng)
+                    xy = tform.apply(unit_gridpoints(shape, mode="xy")[shuffle])
+                    out = estimate_grid_orientation(xy, shape, AffineTransform)
+                    for name in ("matrix", "translation"):
+                        numpy.testing.assert_array_almost_equal(
+                            getattr(tform, name), getattr(out, name)
+                        )
```
Comment (tepals) [inline src/odemis/util/test/registration_test.py:381]: This is hard to read. I would remove the for loop and write it like:
```
numpy.testing.assert_array_almost_equal(tform.matrix, out.matrix)
numpy.testing.assert_array_almost_equal(tform.translation, out.translation)
```

File: src/odemis/util/test/registration_test.py [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+                with self.subTest(shape=shape, i=i):
+                    n, m = shape
+                    shuffle = self._rng.permutation(n * m)
+                    tform = random_affine_transform(self._rng)
+                    xy = tform.apply(unit_gridpoints(shape, mode="xy")[shuffle])[:-1]
+                    out = estimate_grid_orientation(xy, shape, AffineTransform)
+                    for name in ("matrix", "translation"):
+                        numpy.testing.assert_array_almost_equal(
+                            getattr(tform, name), getattr(out, name)
+                        )
```
Comment (tepals) [inline src/odemis/util/test/registration_test.py:400]: same here, this is also hard to read, I would rewrite it to:

```
numpy.testing.assert_array_almost_equal(tform.matrix, out.matrix)
numpy.testing.assert_array_almost_equal(tform.translation, out.translation)
```

File: src/odemis/util/registration.py [hunk 17]
```diff
@@ -0,0 +1,464 @@
...
+    threshold_rel : float, optional
+        If provided, apply a threshold on the minimum intensity of peaks,
+        calculated as `max(image) * threshold_rel`.
+    num_spots : int, optional
+        Maximum number of spots. When the number of spots exceeds `num_spots`,
+        return `num_spots` peaks based on highest spot intensity. Will use
+        `num_spots = shape[0] * shape[1]` as default when set to `None`. Set
+        `num_spots = 0` to not impose a maximum. Note that this behavior is
+        different from odemis.util.spot.find_spot_position().
+
```
Comment (tepals) [inline src/odemis/util/registration.py:456]: can you add an explanation about what it returns? And is the returned transform a square grid, or can it contain shear etc?

**Review by tepals — COMMENTED** [review]
a few more comments

File: src/odemis/util/test/registration_test.py (hunk 15 above)
Comment (effting) [reply src/odemis/util/test/registration_test.py]: done

File: src/odemis/util/test/registration_test.py (hunk 16 above)
Comment (effting) [reply src/odemis/util/test/registration_test.py]: done

File: src/odemis/util/__init__.py [hunk 18]
```diff
@@ -24,28 +24,43 @@
...
 import time
-import weakref
 import types
+import weakref
+from concurrent.futures import CancelledError
+from functools import wraps
+from typing import Iterable, Tuple, TypeVar
+
+import numpy
+from decorator import decorator
```
Comment (pieleric) [inline src/odemis/util/__init__.py:46]: Ideally, minimize the changes to the import part which is not affecting any behaviour. It's much less clear which import you have added/removed.

File: src/odemis/util/graph.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/util/graph.py:50]: Please keep in mind that PEP8 should not be strictly applied. Judgement should be used for odd cases. That's what PEP8 also defines!

**Review by pieleric — CHANGES_REQUESTED** [review]
Looks alright. I just have some small comments on the style.
However, my main question is: what is this code for? There is no user. It seems that it brings a single function directly useful: estimate_grid_orientation_from_img(), right? Everything else is dependency function, correct?
Could you clarify which code is going to use this new function? Or add a commit which convert some of the code current code to use this "improved version" of FindGridSpots?

File: src/odemis/util/__init__.py (hunk 18 above)
Comment (effting) [reply src/odemis/util/__init__.py:46]: The code got passed through `isort` as a pre-commit. Do you want me to revert to the original ordering?

File: src/odemis/util/graph.py [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+        visited: List[Set[int]] = [set() for _ in range(n)]
+        for s in vertices:
+            for t in self.data[s]:
+                if index[s] < index[t]:
+                    for v in visited[s].intersection(visited[t]):
+                        yield tuple(sorted((v, s, t)))
+                    visited[t].add(s)
+
+    def remove_triangles(self) -> None:
+        """Remove all triangles (3-cycles) from the graph."""
```
[the remark this answers is not in the document]
Comment (effting) [reply src/odemis/util/graph.py]: Updated docstring

File: src/odemis/util/__init__.py (hunk 18 above)
Comment (pieleric) [reply src/odemis/util/__init__.py:46]: Yes please revert. Now that there should be very little changes, I think it's much easier to read your changes without the resorting.
I don't recommend having some sort of automatic code rewriting before your commit, as it's hard to control. I'd recommend having it part of your IDE.

**Review by pieleric — CHANGES_REQUESTED** [review]
Yes, adding a commit with the update to the spot-grid script showing how this can now be used would be nice. Maybe also immediately too to `src/odemis/acq/fastem.py`. This way all the users in Odemis would be converted.
Eventually FindGridSpots be obsoleted, and we'll be able to drop it, right?

**Pull request merged**